File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(heights) {
  let leftWallIndex = 0;
  let rightWallIndex = heights.length - 1;
  let maxWater = 0;

  while (leftWallIndex <= rightWallIndex) {
    const width = rightWallIndex - leftWallIndex;
    const height = Math.min(heights[leftWallIndex], heights[rightWallIndex]);
    const waterArea = height * width;
     
    if (waterArea >  maxWater) {
      maxWater = waterArea;
    }

    if (heights[leftWallIndex] >= heights[rightWallIndex]) {
      rightWallIndex--;
    } else {
      leftWallIndex++;
    }
  }

  return maxWater;
};

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left_wall = 0
        right_wall = len(height) - 1
        max_water = 0

        while left_wall < right_wall:
            if height[left_wall] > height[right_wall]:
                max_water = max((right_wall - left_wall) * height[right_wall], max_water)
                right_wall -= 1
            else:
                max_water = max((right_wall - left_wall) * height[left_wall], max_water)
                left_wall += 1
        
        return max_water